pyteal/ast/abi/uint.py

```diff
@@ -0,0 +1,195 @@
+from typing import Union, cast
+from abc import abstractmethod
+
+from ...types import TealType
+from ..expr import Expr
+from ..seq import Seq
+from ..assert_ import Assert
+from ..substring import Suffix
+from ..int import Int
+from ..bytes import Bytes
+from ..unaryexpr import Itob, Not
+from ..binaryexpr import GetByte, ExtractUint16, ExtractUint32, ExtractUint64
+from ..ternaryexpr import SetByte
+from .type import Type
+
+NUM_BITS_IN_BYTE = 8
+
+
+class Uint(Type):
+    def __init__(self, bit_size: int) -> None:
+        valueType = TealType.uint64 if bit_size <= 64 else TealType.bytes
+        super().__init__(valueType)
+        self.bit_size = bit_size
+
+    def has_same_type_as(self, other: Type) -> bool:
+        return isinstance(other, Uint) and self.bit_size == other.bit_size
+
+    def is_dynamic(self) -> bool:
+        return False
+
+    def bits(self) -> int:
+        return self.bit_size
+
+    def byte_length_static(self) -> int:
+        return self.bit_size // NUM_BITS_IN_BYTE
+
+    def get(self) -> Expr:
+        return self.stored_value.load()
+
+    @abstractmethod
+    def set(self, value: Union[int, Expr]) -> Expr:
+        pass
+
+    def __str__(self) -> str:
+        return "uint{}".format(self.bit_size)
+
+
+class Uint8(Uint):
+    def __init__(
+        self,
+    ) -> None:
+        super().__init__(8)
+
+    def new_instance(self) -> "Uint8":
+        return Uint8()
+
+    def set(self, value: Union[int, Expr]) -> Expr:
+        if type(value) is int:
+            if value >= 2 ** 8:
+                raise ValueError("Value exceeds Uint8 maximum: {}".format(value))
+            value = Int(value)
+        # TODO: check dynamic value bounds?
+        return self.stored_value.store(cast(Expr, value))
+
+    def decode(
+        self,
+        encoded: Expr,
+        *,
+        startIndex: Expr = None,
+        endIndex: Expr = None,
+        length: Expr = None
+    ) -> Expr:
+        if startIndex is None:
+            startIndex = Int(0)
+        return self.set(GetByte(encoded, startIndex))
+
+    def encode(self) -> Expr:
+        # value might exceed a uint16, need to check at runtime
+        return SetByte(Bytes(b"\x00"), Int(0), self.get())
+
+
+Uint8.__module__ = "pyteal"
+
+Byte = Uint8
+
+
+class Uint16(Uint):
+    def __init__(
+        self,
+    ) -> None:
+        super().__init__(16)
+
+    def new_instance(self) -> "Uint16":
+        return Uint16()
+
+    def set(self, value: Union[int, Expr]) -> Expr:
+        if type(value) is int:
+            if value >= 2 ** 16:
+                raise ValueError("Value exceeds Uint16 maximum: {}".format(value))
+            value = Int(value)
+        # TODO: check dynamic value bounds?
```

Comment:
nit: is it possible (or worth it) to try something like
```python
Seq(
    Assert(value < Int(2 ** 16)),
    self.stored_value.store(cast(Expr, value)
)
```
for the stack variable (which is the dynamic value bound here)?

Reply:
Yes, that's the kind of check this comment is referencing. I think for correctness we want this, but for efficiency we would also want an optimization step that removes as much of these checks as possible if they are not necessary.

Reply:
I see, the question is more on if it "worth" adding these checks on stack variables then.

Reply:
I think I've decides, yes, this check should be here. A later PR can attempt to optimize it